lib/core/test/unit/Cardano/Wallet/Primitive/AddressDiscovery/DelegationSpec.hs

```diff
@@ -0,0 +1,329 @@
+isConsecutiveRange [_] = True
+isConsecutiveRange [] = True
+isConsecutiveRange (a:b:t)
+    | b == a + 1 = isConsecutiveRange (b:t)
+    | otherwise  = False
+
+
+--
+-- Mock Stake Keys
+--
+
+-- | Mock key type for testing.
+--
+-- FIXME: We should do /some/ testing with @ShelleyKey@ though.
+newtype StakeKey' (depth :: Depth) key = StakeKey' Word
+    deriving newtype (Eq, Enum, Ord, Show, Bounded)
+
+type StakeKey = StakeKey' 'AddressK XPub
+
+instance ToRewardAccount StakeKey' where
+    toRewardAccount (StakeKey' i) = RewardAccount . B8.pack $ show i
+    someRewardAccount = error "todo: this should take ix, presumably?"
+
+instance HardDerivation StakeKey' where
+    type AddressIndexDerivationType StakeKey' = 'Soft
+    deriveAccountPrivateKey _ _ _ = error "todo"
+    deriveAddressPrivateKey _ _ _ _ = error "todo"
+
+instance SoftDerivation StakeKey' where
+    deriveAddressPublicKey _acc _role i = StakeKey' $ toEnum $ fromEnum i
+
+--
+-- Mock chain of delegation certificates
+--
+
+newtype Tx = Tx { unTx :: [Cert] }
+    deriving (Eq, Generic)
+    deriving Show via (Quiet Tx)
+
+instance Arbitrary RewardAccount where
+    arbitrary = toRewardAccount @StakeKey' <$> arbitrary
+
+instance Arbitrary Cert where
+    arbitrary = genericArbitrary
+    shrink = genericShrink
+
+instance Arbitrary Tx where
+    arbitrary = genericArbitrary
+    shrink = genericShrink
+
+instance Arbitrary (StakeKey' depth key) where
+    arbitrary = StakeKey' <$> arbitrary
+
+--
+-- Mock chain data
+--
+
+data Cmd
+    = CmdSetPortfolioOf Int
+    | CmdAdversarialReg RewardAccount
+      -- ^ Someone could pay 2 ada to (re-)register your stake key. Your wallet
+      -- shouldn't be affected negatively from it.
+    deriving (Generic, Eq)
+
+isAdversarial :: Cmd -> Bool
+isAdversarial (CmdSetPortfolioOf _) = False
+isAdversarial (CmdAdversarialReg _) = True
+
+instance Show Cmd where
+    show (CmdSetPortfolioOf n) = "CmdSetPortfolioOf " <> show n
+    show (CmdAdversarialReg (RewardAccount a)) = "CmdAdversarialReg " <> B8.unpack a
+
+instance Arbitrary Cmd where
+    -- We don't want to generate too many adversarial registrations (we don't
+    -- expect them to happen in reality), but at least some, and enough to cause
+    -- consistent failures if something is wrong.
+    arbitrary = frequency
+        [ (98, CmdSetPortfolioOf . getNonNegative <$> arbitrary)
+        , (0, CmdAdversarialReg <$> arbitrary) -- TODO: Re-enable
+        ]
+    shrink = genericShrink
+
+cmdsFromChain :: [Tx] -> [Cmd]
+cmdsFromChain =
+    map (CmdSetPortfolioOf . length . activeKeys)
+    . scanl (flip applyTx) (initialDelegationState accK)
+
+chainFromCmds :: [Cmd] -> [Tx]
+chainFromCmds goals = map Tx . filter (not . null) $ reverse $ go 0 [] [] goals
+  where
+    -- NOTE: For convenience we treat ix=-1 as no keys, as ix=0 means the first
+    -- key (at index 0) is registered and delegating.
+    go i chain tx (cmd:rest) = case cmd of
+        CmdSetPortfolioOf n
```

Comment:
Just realized that this will never generate delegation certificates on their own, which should be pretty important to test.